interface/ceed-basis.c

```diff
@@ -1393,6 +1391,203 @@ int CeedBasisApply(CeedBasis basis, CeedInt num_elem, CeedTransposeMode t_mode,
+        }
+      }
+
+      // Inverse of coefficient matrix
+      CeedCall(CeedCalloc(Q_1d * Q_1d, &basis->legendre_coeffs_1d));
+      CeedCall(CeedCalloc(Q_1d * Q_1d, &Q_inv));
+      CeedCall(CeedCalloc(Q_1d, &tau));
+      // -- QR Factorization, C = Q R
+      CeedCall(CeedQRFactorization(basis->ceed, C, tau, Q_1d, Q_1d));
+      // -- Cinv = Q^T Rinv
+      for (CeedInt i = 0; i < Q_1d; i++) Q_inv[i * Q_1d + i] = 1.0;
+      CeedCall(CeedHouseholderApplyQ(Q_inv, C, tau, CEED_TRANSPOSE, Q_1d, Q_1d, Q_1d, Q_1d, 1));
+      for (CeedInt j = 0; j < Q_1d; j++) {  // Column j
+        basis->legendre_coeffs_1d[j + Q_1d * (Q_1d - 1)] = Q_inv[j + Q_1d * (Q_1d - 1)] / C[Q_1d * Q_1d - 1];
+        for (CeedInt i = Q_1d - 2; i >= 0; i--) {  // Row i
+          basis->legendre_coeffs_1d[j + Q_1d * i] = Q_inv[j + Q_1d * i];
+          for (CeedInt k = i + 1; k < Q_1d; k++) {
+            basis->legendre_coeffs_1d[j + Q_1d * i] -= C[k + Q_1d * i] * basis->legendre_coeffs_1d[j + Q_1d * k];
+          }
+          basis->legendre_coeffs_1d[j + Q_1d * i] /= C[i + Q_1d * i];
+        }
+      }
+      // Cleanup
+      CeedCall(CeedFree(&Q_inv));
+      CeedCall(CeedFree(&tau));
+      CeedCall(CeedFree(&C));
+    } else {
+      // ToDo: Add non-tensor implementation
+      return CeedError(basis->ceed, CEED_ERROR_UNSUPPORTED, "CeedBasisApplyAtPoints not supported for non-tensor basis");
+    }
+  }
+
+  // Basis evaluation
+  CeedVector        q           = NULL;
+  const CeedScalar *x_ref_array = NULL;
+
+  CeedCall(CeedVectorGetArrayRead(x_ref, CEED_MEM_HOST, &x_ref_array));
+  switch (t_mode) {
+    case CEED_NOTRANSPOSE: {
+      // Nodes to arbitrary points
+      CeedInt           Q, Q_comp;
+      CeedScalar       *v_array = NULL;
+      const CeedScalar *q_array = NULL;
+
+      // -- Interpolate to quadrature points
+      CeedCall(CeedBasisGetNumQuadraturePoints(basis, &Q));
+      CeedCall(CeedBasisGetNumQuadratureComponents(basis, CEED_EVAL_INTERP, &Q_comp));
+      CeedCall(CeedVectorCreate(basis->ceed, Q * Q_comp, &q));
+      CeedCall(CeedBasisApply(basis, 1, CEED_NOTRANSPOSE, CEED_EVAL_INTERP, u, q));
+
+      // -- Map to Legendre coefficients
+      CeedInt    pre = Q * Q_comp, post = 1;
+      CeedScalar coeffs[Q * Q_comp];
+
+      for (CeedInt i = 0; i < Q * Q_comp; i++) coeffs[i] = 0.0;
+      if (basis->contract) {
+        CeedScalar tmp[2][Q * Q_comp];
+
+        CeedCall(CeedVectorGetArrayRead(q, CEED_MEM_HOST, &q_array));
+        for (CeedInt d = 0; d < dim; d++) {
+          CeedCall(CeedTensorContractApply(basis->contract, pre, Q, post, Q, basis->legendre_coeffs_1d, CEED_NOTRANSPOSE, d > 0,
+                                           d == 0 ? q_array : tmp[d % 2], d == dim - 1 ? coeffs : tmp[(d + 1) % 2]));
+          pre /= Q;
+          post *= Q;
+        }
+        CeedCall(CeedVectorRestoreArrayRead(q, &q_array));
+      }
```

Comment:
I should replace this with a "basis" that just interpolates since I'm repeating logic here

Reply:
@jedbrown feel free to fiddle with this branch while I'm gone if you are so inclined.